Modular Side Menu

yujinyuz opened the pull request:
Please ensure you are familiar with our Pull Request Description expectations described here: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/
### What?

Modularize side menu items.

Closes #1437

### Why?

These changes will let the users customize their own decks

### How?

This includes a migration for adding new fields to the MenuItem model, changes to the menu item delete view. A method under the model manager is also added for retrieving or creating the default menu items.

### Testing?

I've added tests for creating default side menu items, and also prevents users from deleting a default menu item.

### Screenshots (if front end is affected)


https://github.com/bytedeck/bytedeck/assets/10972027/fefbb8ed-4628-4b8b-a459-c1a69351a99e



### Anything Else?

- I also modified the update menu item to include a preview for the fa icon
- I added a new profile url `/profiles/own/` because the current one requires a primary key and I don't think you can include that within the `url` of the MenuItem
- 

### Review request
@tylerecouture

tylerecouture reviewed (changes requested):
I haven't looked at code, but I fetched the PR to test it out live.  This is awesome!  A few tweaks requested.  We want to minimize the effect on existing users (they should see no change, other than an option to edit the list)

1. The top group doesn't seem to have the correct bootstrap styles (missing bottom border)
2. swap Students and Quest Approvals buttons so Quest Approvals is on the bottom (as before)
![image](https://github.com/bytedeck/bytedeck/assets/10604391/677edc13-4015-4030-91c6-ef9d620e1b9d)

When I did the migration, I lost some of the menu items that used to be there.  Those items should be the defaults on a new deck and existing deck.  See before on the right and after on the left:
![image](https://github.com/bytedeck/bytedeck/assets/10604391/9a022a87-4c0f-4aaa-aa84-b9f9f4cde415)

Please remove the margin above the "Edit this list" link so it is more associated with the middle button group that can be customized and change text to "Customize this menu".
![image](https://github.com/bytedeck/bytedeck/assets/10604391/a7beffac-5cfb-4045-a9ab-c92730384dab)


To differentiate the list views, please title the new one "Side Menu Items"

yujinyuz commented:
@tylerecouture Can you try opening it in an incognito window? I already changed the css and experienced the same thing where it still had no borders. Not sure how to properly fix the cached css issue on existing decks though.

Updated PR to reflect with the requested changes. Can you test again? Thanks!

tylerecouture commented:
> @tylerecouture Can you try opening it in an incognito window? I already changed the css and experienced the same thing where it still had no borders. Not sure how to properly fix the cached css issue on existing decks though.
> 
> Updated PR to reflect with the requested changes. Can you test again? Thanks!

Try cache busting by incrementing CSS file version in the url after the ? In the (header?) template.

yujinyuz commented:
@tylerecouture thanks! didn't know about this. updated the version and it should work now

tylerecouture commented on src/profile_manager/urls.py:
```diff
@@ -23,6 +23,7 @@
     url(r'^(?P<profile_id>[0-9]+)/comment_ban_toggle/$', views.comment_ban_toggle, name='comment_ban_toggle'),
     url(r'^(?P<profile_id>[0-9]+)/comment_ban/$', views.comment_ban, name='comment_ban'),
     url(r'^(?P<pk>[0-9]+)/edit/$', views.ProfileUpdate.as_view(), name='profile_update'),
+    url(r'^own/$', views.ProfileDetailOwn.as_view(), name='profile_detail_own'),
```
Can you add basic access tests for this new view? For anonymous, random authenticated (verifying it is their own), and staff?

yujinyuz replied:
added

tylerecouture commented on src/profile_manager/views.py:
```diff
@@ -163,16 +163,28 @@ class ProfileDetail(NonPublicOnlyViewMixin, DetailView):
+        if pk is not None:
+            profile = get_object_or_404(Profile, pk=self.kwargs.get('pk'))
+        else:
+            profile = self.request.user.profile
```
L180 Is not covered by a test: `profile = self.request.user.profile`

tylerecouture reviewed (changes requested):
@yujinyuz Hopefully last series of tweaks required!

yujinyuz commented on src/utilities/templates/utilities/menuitem_form.html:
```diff
@@ -12,9 +12,22 @@
+{% endblock %}
+
+{% block js %}
+<script>
```
This is just a nice little preview added when typing a font awesome class

tylerecouture replied:
oh!  Very cool!

tylerecouture replied:
Very cool!

yujinyuz commented on src/utilities/templates/utilities/side_menu_items.html:
```diff
@@ -0,0 +1,12 @@
+  </a>
+{% endfor %}
+</div>
+<a href="{% url 'utilities:menu_items' %}?is_side_menu=true"><i class="fa fa-fw fa-pencil"></i>&nbsp;&nbsp;Customize this menu</a>
```
Can't deal with this atm. Not sure if I can apply an immediate fix for this. Will deal with this on the weekend

tylerecouture commented on src/profile_manager/tests/test_views.py:
```diff
@@ -176,6 +179,23 @@ def test_profile_detail__student_marks_button(self):
         response = self.client.get(reverse('profiles:profile_detail', args=[s_pk]))
         self.assertContains(response, 'View your Mark Calculations')
 
+    def test_profile_own__correct_displayed_text(self):
```
Now that this view is serving both profile_own and other profile_detail, there should also be a test to confirm profile_detail actually returns the correct profile (and not own).

Currently we are only checking for 200 response.

tylerecouture commented on src/profile_manager/views.py:
```diff
@@ -163,16 +163,28 @@ class ProfileDetail(NonPublicOnlyViewMixin, DetailView):
 
+    def get_object(self):
+        pk = self.kwargs.get('pk')
+        profile = self.request.user.profile
```
Why is `profile = self.request.user.profile` added here if ProfileDetailOwn has it's own view?

yujinyuz replied:
I was going through the code again and I kinda remember now that the sole reason I added this was just so profile doesn't get set to `None`.

I removed this and replaced it with raising an Http404 instead.

tylerecouture commented on src/utilities/templates/utilities/side_menu_items.html:
```diff
@@ -0,0 +1,11 @@
+  </a>
+{% endfor %}
+</div>
+<a href="{% url 'utilities:menu_items' %}?is_side_menu=true" class="hidden-xs" style="text-decoration: none; color: #777;"><i class="fa fa-fw fa-pencil"></i>&nbsp;&nbsp;Customize this menu</a>
```
Why did you opt for using get parameters?  This feels clunky to me and I think makes it difficult for future developers, since we don't use get parameters anywhere else in this project.  Even here for example, it's a hard coded addition that is prone to easy breakage and goes against the whole django url design philosophy.  This should be a view with it's own name.

tylerecouture commented on src/utilities/templates/utilities/side_menu_items_navbar.html:
```diff
@@ -0,0 +1,72 @@
+            </a>
+        </li>
+        {% endif %}
+        {% endif %}
```
Can you tab this code so the if blocks aren't aligned?

tylerecouture commented on src/utilities/templatetags/utility_tags.py:
```diff
@@ -58,10 +58,22 @@ def group_name():
+@register.inclusion_tag('utilities/side_menu_items_navbar.html', takes_context=True)
+def side_menu_list_navbar(context):
+    side_menu_items = MenuItem.objects.get_side_menu_items().get_visible_items()
+    return {'side_menu_items': side_menu_items, 'request': context.request if 'request' in context else None}
```
These methods don't have any doc strings. What's the purpose behind this check for `request`?

tylerecouture replied:
Docstrings should make it more clear what the difference is between these two tags.

tylerecouture commented:
@yujinyuz can you reply to each of my comments and mark resolved if they are?
